controlplane/internal/imageregistry/image_registry.go

```diff
@@ -93,91 +119,73 @@ func CreateImageRegistryConfigmap(imageRegistry *corev1.ConfigMap, namespace str
 // It will convert it to an ImageDigestMirrorSet CR and/or an ImageTagMirrorSet.
 // It will return these as marshalled JSON strings, and it will return a string list of insecure
 // mirror registries (if they exist)
-func getImageRegistries(registry string) (string, string, []string, error) {
+func getImageRegistries(registry string, mirrorConfigs *MirrorConfigs) error {
 	// Parse toml and add mirror registry to image digest mirror set
 	tomlTree, err := toml.Load(registry)
 	if err != nil {
-		return "", "", nil, err
+		return errors.Wrap(err, "failed to load value of registries.conf into toml tree; incorrectly formatted toml")
 	}
 	registriesTree, ok := tomlTree.Get("registry").([]*toml.Tree)
 	if !ok {
-		return "", "", nil, fmt.Errorf("failed to find registry in toml tree")
+		return fmt.Errorf("failed to find registry key in toml tree")
 	}
 
-	idmsMirrors := make([]configv1.ImageDigestMirrors, 0)
-	itmsMirrors := make([]configv1.ImageTagMirrors, 0)
-	insecureRegistries := make([]string, 0)
-
 	// Add each registry and its mirrors as either an image digest mirror or an image tag mirror
 	for _, registry := range registriesTree {
 		source, sourceExists := registry.Get("location").(string)
 		mirrorTrees, mirrorExists := registry.Get("mirror").([]*toml.Tree)
-		if !sourceExists || !mirrorExists {
-			continue
+		if !sourceExists {
+			return fmt.Errorf("failed to parse image registry toml, missing registry location")
+		}
+		if !mirrorExists {
+			return fmt.Errorf("failed to parse image registry toml, missing registry.mirror in registries.conf")
+		}
+		if err := parseMirrorRegistries(mirrorConfigs, mirrorTrees, source); err != nil {
```

Comment:
Why not returning only an object here with ImageTagMirrors, ImageDigestMirrors and insecure registry, and move the merge/addition logic outside?
The benefit I see with this is that we would use a more functional approach, no need to provide any data to the function and it will always return the same value with the same input, no matter what we have been parsing before.

Reply:
Ok modified it!